File: Trees/Interviewbit/pathsum.py
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
bool checkSum(TreeNode* root, int sum){
    if(root == NULL){
        return 0;
    }
    
    if(root->val == sum && root->left == NULL && root->right == NULL){
        return 1;
    }
    
    return (checkSum(root->left, sum - root->val) || checkSum(root->right, sum - root->val));
}

 
 
int Solution::hasPathSum(TreeNode* A, int B) {
    
    return checkSum(A,B);
    
    
}

File: Trees/LeetCode/pathsum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool solve(TreeNode* root, int sum){
        
        if(root==NULL){
            return false;
        }
        
       if(root->val==sum && root->left==NULL && root->right==NULL){
           return true;
       }
        
        return solve(root->left, sum-root->val) || solve(root->right, sum-root->val);
        
        
        
        
    }
    
    bool hasPathSum(TreeNode* root, int sum) {
        
       return solve(root, sum);
        
        
    }
};
